tls/extensions/s2n_server_key_share.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/extensions/s2n_server_key_share.h"
+
+#include "tls/s2n_client_extensions.h"
+#include "utils/s2n_safety.h"
+
+/*
+ * Calculate the data length for Server Key Share extension
+ *
+ * negotiated_curve should be selected in server_ecc_params.
+ */
+int s2n_extensions_server_key_share_send_size(struct s2n_connection *conn)
+{
+    const struct s2n_ecc_named_curve* curve = conn->secure.server_ecc_params.negotiated_curve;
+
+    if (curve == NULL) {
+        return 0;
```

Comment:
I *think* this should be an error. We'll underflow here https://github.com/awslabs/s2n/pull/1151/files#diff-289dd68a4731135401130dee225072f6R54 if 0 is actually a valid value to return. 

I'm not sure if we were trying to account for key exchange modes that don't need to send a key_share here(we don't support any of those, yet).

Reply:
`s2n_extensions_server_key_share_send_size()` and `s2n_extensions_server_key_share_send()` are both used in `s2n_server_extensions.c`. The current usage of `*_size()` are unguarded in both `s2n_server_extensions` and `s2n_client_extensions` as it's primary function is to calculate a size. The error is currently not caught in `.._size()` as it would be caught in `.._send()`. 

```c
   server_extension_size += s2n_extensions_server_key_share_send_size(conn);
   ...
   s2n_extensions_server_key_share_send(conn); // error is caught and propagated here
```

Do you think it would be acceptable to caught the error in `*_send()` functions or have something like
```
   int extension_size = s2n_extensions_server_key_share_send_size(conn);
   ERROR_IF(extension_size < 0, ERR_NO);
   server_extension_size += extension_size;
   ...
   s2n_extensions_server_key_share_send(); /* should errors still be checked here? */
```
which size calculation code a little unwieldy?

Reply:
I think I prefer your second proposal. Another option is to assume curve not being null is an invariant of calling this size helper similar to what @ttjsu-aws suggested elsewhere.

Reply:
I prefer assume that curve is not null by the time this code path is executed, the checks here is an additional sanity check.